File: JAVA/final-BT-java/src/repositories/DashboardRepository.java
package repositories;

import config.DatabaseConnection;

import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class DashboardRepository {

    private final Connection connection;

    public DashboardRepository() {
        connection = DatabaseConnection
                .getInstance()
                .getConnection();
    }

    public BigDecimal getRevenus(int idUtilisateur) {

        String sql = """
            SELECT COALESCE(SUM(montant),0)
            FROM transaction
            WHERE id_utilisateur=?
            AND type='REVENU'
            AND DATE_TRUNC('month',date_transaction)
            =DATE_TRUNC('month',CURRENT_DATE)
            """;

        try (PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setInt(1,idUtilisateur);

            ResultSet rs = statement.executeQuery();

            if(rs.next())
                return rs.getBigDecimal(1);

        } catch (Exception e) {
            e.printStackTrace();
        }

        return BigDecimal.ZERO;
    }

    public BigDecimal getDepenses(int idUtilisateur){

        String sql="""
            SELECT COALESCE(SUM(montant),0)
            FROM transaction
            WHERE id_utilisateur=?
            AND type='DEPENSE'
            AND DATE_TRUNC('month',date_transaction)
            =DATE_TRUNC('month',CURRENT_DATE)
            """;

        try(PreparedStatement statement=connection.prepareStatement(sql)){

            statement.setInt(1,idUtilisateur);

            ResultSet rs=statement.executeQuery();

            if(rs.next())
                return rs.getBigDecimal(1);

        }catch(Exception e){
            e.printStackTrace();
        }

        return BigDecimal.ZERO;
    }

}

File: PHP/src/Repository/DashboardRepository.php
<?php

namespace App\Repository;

use App\Config\AbstractRepository;
use PDO;

class DashboardRepository extends AbstractRepository
{

    public function getDashboard(int $idUtilisateur): array
{
    $sql = "
    SELECT

        u.solde,

        COALESCE((
            SELECT SUM(montant)
            FROM transaction t
            WHERE t.id_utilisateur = :id
            AND t.type = 'REVENU'
            AND DATE_TRUNC('month', t.date_transaction) =
                DATE_TRUNC('month', CURRENT_DATE)
        ),0) AS revenus,

        COALESCE((
            SELECT SUM(montant)
            FROM transaction t
            WHERE t.id_utilisateur = :id
            AND t.type = 'DEPENSE'
            AND DATE_TRUNC('month', t.date_transaction) =
                DATE_TRUNC('month', CURRENT_DATE)
        ),0) AS depenses,

        COALESCE((
            SELECT COUNT(*)
            FROM budget
            WHERE id_utilisateur = :id
        ),0) AS budgets,

        COALESCE((
            SELECT COUNT(*)
            FROM transaction
            WHERE id_utilisateur = :id
        ),0) AS transactions

    FROM utilisateur u
    WHERE u.id_utilisateur = :id
    ";
    $statement = $this->database->prepare($sql);
    $statement->execute(["id" => $idUtilisateur]);
    return $statement->fetch(PDO::FETCH_ASSOC);
}
}
